plugins/net_plugin/net_plugin.cpp

```diff
@@ -3277,36 +3271,66 @@ namespace eosio {
       close( retry ); // reconnect if wrong_version
    }
 
+   // some clients before leap 5.0 provided microsecond epoch instead of nanosecond epoch
+   tstamp normalize_epoch_to_ns(tstamp x) {
+      int digits = 1;
+      tstamp v = x;
+      while (v/=10)
+         ++digits;
+      // 168 1685 9763 7880 7944 >= 19 is nanoseconds
+      //     1685 9763 7880 7944 >= 16 is microseconds
+      //        1 6859 7637 8807 >= 13 is milliseconds
+      //            16 8597 6378 >= 10 is seconds
+      if (digits >= 19)
+         return x;
+      if (digits >= 16)
+         return x*1000;
+      if (digits >= 13)
+         return x*1000*1000;
+      if (digits >= 10)
+         return x*1000*1000*1000;
+      return 0; // unknown or is zero
+   }
+
    void connection::handle_message( const time_message& msg ) {
-      peer_ilog( this, "received time_message" );
+      peer_dlog( this, "received time_message: ${t}", ("t", msg) );
 
-      /* We've already lost however many microseconds it took to dispatch
-       * the message, but it can't be helped.
-       */
+      // We've already lost however many microseconds it took to dispatch the message, but it can't be helped.
       msg.dst = get_time();
 
       // If the transmit timestamp is zero, the peer is horribly broken.
       if(msg.xmt == 0)
-         return;                 /* invalid timestamp */
+         return; // invalid timestamp
 
-      if(msg.xmt == xmt)
-         return;                 /* duplicate packet */
+      auto msg_xmt = normalize_epoch_to_ns(msg.xmt);
+      auto msg_org = normalize_epoch_to_ns(msg.org);
 
-      xmt = msg.xmt;
-      rec = msg.rec;
-      dst = msg.dst;
+      if (msg_org != 0 && msg_org == normalize_epoch_to_ns(org)) {
+         auto latency = msg.dst - msg_org;
+         peer_dlog(this, "send_time latency ${l}us", ("l", latency/2/1000));
+         net_latency_ns = latency/2;
+      }
+
+      if (msg_xmt == xmt)
+         return; // duplicate packet
+
+      xmt = msg_xmt;
+      rec = normalize_epoch_to_ns(msg.rec);
+      dst = msg.dst; // already normalized
 
       if( msg.org == 0 ) {
          send_time( msg );
          return;  // We don't have enough data to perform the calculation yet.
       }
 
-      double offset = (double(rec - org) + double(msg.xmt - dst)) / 2;
-      double NsecPerUsec{1000};
+      if (org != 0) {
+         int64_t offset = (double(rec - org) + double(msg_xmt - dst)) / 2.0;
```

Comment:
`msg_xmt - dst` is supposed to be negative, I think. Luckily the tstamp seems to be holding a signed value.

It looks like we assume that the network latency is the same in both directions, which I think may be optimistic for a single round trip. Maybe we should average the offset over time, so it would keep getting more accurate as we exchange messages?

Reply:
The offset is not used outside logging it. I just kept what was already there.